microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/MicroProfileConfigPropertyDiagnosticsParticipant.java

```diff
@@ -0,0 +1,137 @@
+import org.eclipse.jdt.core.dom.FieldDeclaration;
+import org.eclipse.jdt.core.dom.ITypeBinding;
+import org.eclipse.jdt.core.dom.NormalAnnotation;
+import org.eclipse.jdt.core.dom.StringLiteral;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.DiagnosticSeverity;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4mp.jdt.core.MicroProfileConfigConstants;
+import org.eclipse.lsp4mp.jdt.core.java.diagnostics.IJavaDiagnosticsParticipant;
+import org.eclipse.lsp4mp.jdt.core.java.diagnostics.JavaDiagnosticsContext;
+import org.eclipse.lsp4mp.jdt.core.utils.AnnotationUtils;
+import org.eclipse.lsp4mp.jdt.internal.metrics.MicroProfileMetricsConstants;
+
+/**
+ * Validate "defaultValue" attribute of @ConfigProperty and generate
+ * diagnostics if "defaultValue" cannot be represented by the given
+ * field type.
+ *
+ * See https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/converters.asciidoc
+ * for more details on default converters.
+ */
+public class MicroProfileConfigPropertyDiagnosticsParticipant implements IJavaDiagnosticsParticipant {
+
+	@Override
+	public List<Diagnostic> collectDiagnostics(JavaDiagnosticsContext context, IProgressMonitor monitor)
+			throws CoreException {
+		List<Diagnostic> diagnostics = new ArrayList<>();
+		validateClass(diagnostics, context, monitor);
+		return diagnostics;
+	}
+
+	private void validateClass(List<Diagnostic> diagnostics, JavaDiagnosticsContext context,
+			IProgressMonitor monitor) {
+		CompilationUnit ast = context.getASTRoot();
+		ast.accept(new ConfigPropertiesAnnotationValidator(diagnostics, context));
+	}
+
+	public class ConfigPropertiesAnnotationValidator extends ASTVisitor {
+		private List<Diagnostic> diagnostics;
+		private JavaDiagnosticsContext context;
+
+		public ConfigPropertiesAnnotationValidator(List<Diagnostic> diagnostics, JavaDiagnosticsContext context) {
+			this.diagnostics = diagnostics;
+			this.context = context;
+		}
+
+		@Override
+		public boolean visit(NormalAnnotation node) {
+			if (AnnotationUtils.isMatchAnnotation(node, MicroProfileConfigConstants.CONFIG_PROPERTY_ANNOTATION)) {
+				try {
+					Expression defValueExpr = AnnotationUtils.getAnnotationMemberValueExpression(node, MicroProfileConfigConstants.CONFIG_PROPERTY_ANNOTATION_DEFAULT_VALUE);
+					String defValue = ((StringLiteral) defValueExpr).getLiteralValue();
+					ASTNode parent = node.getParent();
+					if (parent instanceof FieldDeclaration) {
+						FieldDeclaration field = (FieldDeclaration) parent;
+						ITypeBinding fieldBinding = field.getType().resolveBinding();
+						if (fieldBinding != null && !isAssignable(fieldBinding, defValue)) {
+							Range range = context.getUtils().toRange(context.getTypeRoot(), defValueExpr.getStartPosition(), defValueExpr.getLength());
+							diagnostics.add(new Diagnostic(range,
+									"'" + defValue + "'" + " does not match the expected type of '" + fieldBinding.getName() + "'.",
```

Comment:
As discussed, I should probably match this error message to the one used in the application properties file for the mismatched types.

Reply:
Keep like this for the moment because I find your error message better than properties file:

. In application.properties : `Type mismatch: int expected`
* In Java : `'xx' does not match the expected type of 'int'`

I think it's better and application.properties should follow this message.

Reply:
Bah, I just changed it. Let me revert. It's an easy enough change.

Reply:
ok no problem.